### DISCUSSION TITLE
Nerd Fonts 2.2.0 / 2.2.1 / 2.2.2 Release

### DISCUSSION THREAD

**Discussion by Finii** [ryanoasis/nerd-fonts#881] [General]
I tried to identify stuff that should be pulled for 2.2.0 and set the milestones:

https://github.com/ryanoasis/nerd-fonts/pulls?q=+is%3Apr+milestone%3Av2.2.0+

These affect only known bugs in the patched/released fonts and no release-fonts unrelated bugfixes, convenience or feature items.
There are also no 'political' or 'controversial' PRs.

I will try to pull now all these PRs, check the gerenated font set and if all goes well trigger the release. A lot of people struggle with the bugs in 2.1.0 and rely on 'real release' for one reason or another (i.e. because of AUR).
I was very hesitant to pull any PRs, but @ryanoasis is now absent a rather long time and I try to keep to his intentions as good as I can.

Please Ryan, if you do not like that, just tell me and I will stop. But I believe this is in the interest of Nerd Fonts.

**Comment by Finii** [comment]
The first big batch of pulled PRs just went through.
https://github.com/ryanoasis/nerd-fonts/actions/runs/2889080910

Now examining the results
* Release packages 
  * _CI process 'broken', does not upload updated packages_
  * 'Solved' by manually updating 2.2.0-RC release directly
* `patched-fonts` dir 
  * _Lilex missing, solved in #886_
* `font-patcher.zip` :heavy_check_mark: 

Furthermore: Check if we would benefit from also putting #801 on the list. :heavy_check_mark: _included_
All other PRs seem to real feature additions.

**Reply by Finii** [reply]
After pulling #723 regenerating just Caskaydia and (manually) uploading to release page:

![image](https://user-images.githubusercontent.com/16012374/185885791-ad42dda2-90c0-470a-8971-7e260143a2b7.png)

**Comment by Finii** [comment]
Checklist for release:
* ~~Update version in `font-patcher` [1]~~
* **Update version in `package.json`**
* ~~Update other versions [1]~~
* **Commit and push the changes**
* ~~Add release tag to git~~
* ~~more stuff ...~~
* **Release CI is triggered and creates Release**
  * Version in all scripts is updated to `package.json` value
  * Changes are commited back to repo
  * All Fonts are patched
  * Changes are commited back to repo
  * `font-patcher` is Dockerized
  * A new release is created (if not existing)
  * All font archives and `font-patcher` archive are uploaded/updated in the release
* **Fill in detail information in Release's page**

:tada: Seems CI is working as expected with
* #892

----

~~**[1]** There is a script for this `bin/scripts/version-bump.sh`, but that:~~
* ~~does change unrelated stuff like ` # Script Version: 1.0.1`, because the rule is too lax~~
* ~~does never commit the changes it does (when called in the CI)~~

~~So one can use that interactively, but the CI process is broken.~~

**Comment by Finii** [comment]
This PRs release process had to possible problematic behavior to publish the release immediately, i.e. when the first release pack (font) is uploaded. Other fonts will be uploaded one by one later on. People or scripts that check for a new release regularly can be presented with an incomplete release.

The 2.2.0 release had a severe bug anyhow (#895), and in the process of fixing it also the release process has been changed (75ce899201bf9462c133a53cf8664100e5fb0cf7). Now the release is created as draft.

The drawback is that I could not figure out (fast enough) how to un-draft the release. I tried to write to the same release without the draft flag, but that created another same-named but distinct release. This needs still some work.

And then, scoop buckets are automatically updated every hour via cron (https://github.com/matthewjberger/scoop-nerd-fonts), but not mentioned as install method. Not that I have ever used Scoop. Might be a good addition to the docs.

**Reply by Finii** [reply]
Homebrew casks ... https://github.com/Homebrew/homebrew-cask-fonts/pull/6167
AUR https://aur.archlinux.org/packages?O=0&SeB=nd&K=nerd-font&outdated=&SB=p&SO=d&PP=50&submit=Go
Arch https://archlinux.org/packages/?sort=&q=nerd-font&maintainer=&flagged=
Scoop https://github.com/matthewjberger/scoop-nerd-fonts
Chocolatey https://github.com/sumo300/chocolatey-nerd-fonts
Nixpkgs https://github.com/NixOS/nixpkgs/blob/master/pkgs/data/fonts/nerdfonts/update.sh https://search.nixos.org/packages?channel=unstable&query=nerdfonts

Not all mentioned:

![image](https://user-images.githubusercontent.com/16012374/189205466-fc9507da-b334-45eb-94b6-7a889b0c0a87.png)

**Comment by Finii** [comment]
After a lot of thinking and seeing issues, I believe the commit reverted with #764 is indeed a regression of #270.
Nerd Fonts is intended to be used in terminals (mostly) and when it can not, that is a flaw. The original commit #283 fixed a lot of bugs, and they all resurfaced now, after (not with) `v2.1.0` (i.e. already before `v2.2.0-RC`, when 59c45ba4eff539d2a83a7d7004f056b9860d3253 has been merged.

Some backlinks
* #270
* #127 
* #900
* #764
* #746
* #283 
* #731 
* #898 
* #909 
* #442 
* #795

This is a important enough issue to come up with another bugfix release. I guess.

**Reply by Finii** [reply]
I also like to like these external discussions

* https://github.com/microsoft/terminal/issues/11769
* https://github.com/mintty/mintty/issues/1104
* https://github.com/microsoft/terminal/issues/6738
* https://github.com/microsoft/terminal/issues/6864

**Comment by ryanoasis** [comment]
Hey @Finii thank you so much for carrying on. Please forgive my lack of input and involvement of late.

Awesome job on the releases, really appreciate the work and glad things got moving again. I think you made the best decisions you could give complete lack of direction on my part 😢 . It's bittersweet to see the project flourish without me.

I think I just needed a break and it was one of those cases where I had been away so long I wasn't sure how to jump back in and well it's a bit awkward now but oh well 😊 

I'm trying to get back up to speed but there have been quite a lot of changes (which is good for progress bad for me catch up 😆 )

**With that said if there is anything in particular you want my input on please ping it here, otherwise i'll continue to review the releases, discussions, issues, etc. so I can start getting back to helping out here.**

🤞🏻

**Reply by Finii** [reply]
Hej @ryanoasis thanks for the nice words!

And you did leave lots of direction bits, especially with the almost ready `v2.2.0`. At some point I thought: Lets just finish the release, how hard can it be? Well, I learned that it is a full time job :grimacing: 

I hope your break was good, sometimes we all need that. I happily assumed more and more liability for the project, sometimes to the point that my day job has to step back \*cough\*. I'm very very happy you show up, and I will step back any amount; I never 'wanted' the job and you can have back whatever you desire ;) (This also feels awkward to write. Just want to say: I will not be grumpy on any decision you take and you were missed.)

> but there have been quite a lot of changes

From the code point of view I guess the most 'complex' change is support for TTC. They contain multiple fonts in one file, so we needed to decouple files from font-patching (to loop over all contained fonts). That led to some more drastic architectural changes. But I hope it is all documented good enough in the code and if not do a `git blame` because I usually have more things to explain in the commit message.

The hardest decision I took (and everything else can easily be rolled back, but this one is hard once it has been released to the public) was about 'what is the goal of a Nerd Font'. At one point I answered this with: Font for a terminal emulator. This led to this PR
* #764 

Because in principle you can patch the fonts in 3 different ways [2], and we just supply two of them (`Nerd Font` and `Nerd Font Mono`). The candidates where either `Nerd Font Bigicons` or `Nerd Font Proportional`, defining the simple `Nerd Font` to be the other one of the who. See #900 and #951. But because most terminal emulators and other programs assume `Nerd Fonts` to be what `v2.1` was, I reverted the commit that was already in `master` and made `v2.2` the same as what `v2.1` has been. But I think a lot people also want `Nerd Font Proportional` to use the fonts in GUI contexts (good comment by Patrick in #951).
[1]

When one variant is to be added ... I tried to explore if we can drop the `Windows Compat` versions, to minimize size explosion. You already commented on one of that threads.

Some things (changes, PRs) come naturally, but some are more deliberate decisions
* #773 

is one of them. It will grow all font files. Do people really need that? Otoh I virtually 'promised' that change to Aaron in the long standing #365... (he's Cascadia Code).

Well. As you will have seen I prepared a `v2.3` interim release #962 before we break everything with `v3.0` ;-) which is almost ready. I aimed to finish it in October ... \*cough\* I guess it will take one full-time week to close all the bugs on the list.

This comment is too long, sorry.

But let me add one thing that is simmering in the back of my head for quite some time and I have no answer.
At the moment the release candidate 'grows' in `master`, and every commit to master will (probably) rebuild all patched fonts (which I sped up by a factor of 2.5 btw :) ). The simplest change will make the repo grow a lot. But at some comment you point out that there can be release candidate _branches_. That would solve that, they could be deleted after release.
But on the other hand, there are the 3rd party packagers, that take `master` for 'bleeding edge' font packages. The change would break them. "You can just choose between different drawbacks" is what pessimistic-me always say.

So. Welcome back :tada: 


[1] After writing this, I noticed [this](https://github.com/ryanoasis/nerd-fonts/discussions/881#discussioncomment-3581604) comment in this discussion is basically about that decision, includes all my thoughts etc.
[2] #969

_Edit: Add footnote 2_

**Reply by Finii** [reply]
Dear @ryanoasis 

> **With that said if there is anything in particular you want my input on please ping it here, otherwise i'll continue to review the releases, discussions, issues, etc. so I can start getting back to helping out here.**

Finally I set up automatic _Homebrew Cask_ generation (#1008). To more or less automatically create the PR at their repo a mirror of their repo is needed. I saw a similar stub here: https://github.com/NerdFonts/Chocolatey
At the moment the CI works on a fork in my space, but that is of course not preferable.

1. Maybe you can fork https://github.com/Homebrew/homebrew-cask-fonts into the https://github.com/NerdFonts organization (I can not do that)
2. That fork (i.e. organization) needs to have an access token added, so that THIS repo can commit there. (I can not add token to the orga)
3. Put the token into THIS repo as secret (I can not add / edit secrets here)

...

(Maybe such discussion would be more suited for the NerdFonts teams, so please head over there: https://github.com/orgs/NerdFonts/teams/core/discussions/1)
